### DISCUSSION TITLE
"OneToOne composite foreign keys as primary keys" cannot insert

### DISCUSSION THREAD

**Discussion by daniellam258** [mikro-orm/mikro-orm#3842] [Q&A] [answered]
**Describe the bug**
I have a OneToOne relationship between A and B:
1. B references to A.
2. A has composite primary keys.
3. B's composite foreign keys are also its primary keys

When I insert B using QueryBuilder, it's not working

**To Reproduce**
```
@Entity()
export class GamePoolEntity {
  @PrimaryKey()
  contract_address!: string;

  @PrimaryKey()
  chain_id!: number;

  [PrimaryKeyType]?: [string, number];

 @Property()
  rpc_url!: string;

  @Property()
  referral_percents!: number[];

  @Property()
  referral_campaign_id!: number;

  @Property({ defaultRaw: 'now()' })
  created_at!: Date;

  @Property({
    onUpdate: () => new Date(),
    defaultRaw: 'now()',
  })
  updated_at!: Date;
}

@Entity()
export class GamePoolScannerEntity {
  @OneToOne({
    entity: () => GamePoolEntity,
    primary: true,
    wrappedReference: true,
    joinColumns: ['contract_address', 'chain_id'],
  })
  game_pool!: Ref<GamePoolEntity>;

  [PrimaryKeyType]?: [string, number];

  @Property()
  start_block!: number;

  @Property({ nullable: true })
  current_block?: number;

  @Property()
  min_confirmations!: number;

  @Property({ defaultRaw: 'now()' })
  created_at!: Date;

  @Property({
    defaultRaw: 'now()',
    onUpdate: () => new Date(),
  })
  updated_at!: Date;
}

const game_pool = new GamePoolEntity();
await em.createQueryBuilder(GamePoolEntity).insert(game_pool).execute();

const game_pool_scanner = new GamePoolScannerEntity();
game_pool_scanner.game_pool = ref(game_pool);
await em.createQueryBuilder(GamePoolScannerEntity).insert(game_pool_scanner).execute();
```

**Expected behavior**
Can Insert normally, but receive query 
``` insert into "game_pool_scanner_entity" ("contract_address", "min_confirmations", "start_block") values ('{"contract_address":"0x2acf91a4518c5362fcf2a16a164aae87e699dfd5","chain_id":97,"referral_campaign_id":1,"referral_percents":[1000],"rpc_url":"https://nodes.bhostarter.com/testnet/bnb-archive/wss"}', 15, 25279965) returning "contract_address", "chain_id", "created_at", "updated_at"```

It seems that mikro doesn't regconize GamePoolScanner's composite primary keys and try to insert whole GamePool object

**Versions**

| Dependency | Version |
| - | - |
| node | ? |
| typescript | ? |
| mikro-orm | 5.6 |
| your-driver | ? |

**Comment by B4nan** [comment, marked as answer]
You should **not** be putting entity instances into `qb.insert()`, it expects a POJO. Also you are trying to create entity with composite PK without providing the values for it. In fact, all of your properties except the timestamps are required, but you don't provide any value. 

I am quite sure your problem is what I said above, just don't put entity instances into the QB. But if that won't help, I will need to see a complete reproduction.

Is there an actual reason why you want to use QB for this? With UoW/flush it would be much simpler as you could cascade things, single flush would be enough.

edit: this is what the QB expects (and that works, just tried):

```ts
await orm.em.createQueryBuilder(GamePoolScannerEntity).insert({
  game_pool: ['1', 2], // composite PK is represented by a tuple
  start_block: 5,
  min_confirmations: 123,
}).execute();
```

Hmm I see we have a basic support for passing the entities to `qb.insert()` too (https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/query/QueryBuilder.ts#L879-L881).

**Comment by daniellam258** [comment]
Sorry, In the sample code I don't provide the code where I assign values into entity, but yes it's there.

Actually, in the first place I used UoW, but I playaround with QueryBuilder API a little bit because it's just my preferred way to control query.

Also, Along the way I think I encounter another bug with `populate` in `OneToOne` bidirectional. Maybe I will create another discussion.

Tks for that

**Reply by B4nan** [reply]
Sure, ideally try to create executable repros, as in https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

It does not have to be a jest test case, but it should be executable, e.g. it should contain all required data, ORM config and init call, schema generation, etc.

I will try to support what you initially tried, it should be doable, and given we have support for `qb.insert(entity)` and only the composite FK is what fails to work, it is definitely worth fixing.
